### PR TITLE
convert output_device at data_parallel from torch.device to index

### LINKED ISSUE #9984: torch.device and torch.nn.parallel.data_parallel compatibility
## Issue description

``torch.nn.parallel.data_parallel`` does not accept ``torch.device``

## Code example

```
import torch
net = torch.nn.LSTM(10, 10)
torch.nn.parallel.data_parallel(module=net.cuda(), inputs=torch.randn(1, 1, 10).cuda(), output_device=torch.device('cuda')) # Does not work
torch.nn.parallel.data_parallel(module=net.cuda(), inputs=torch.randn(1, 1, 10).cuda(), output_device=torch.device('cuda').index) # Works
```

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/michaelp/.local/lib/python3.5/site-packages/torch/nn/parallel/data_parallel.py", line 169, in data_parallel
    return gather(outputs, output_device, dim)
  File "/home/michaelp/.local/lib/python3.5/site-packages/torch/nn/parallel/scatter_gather.py", line 67, in gather
    return gather_map(outputs)
  File "/home/michaelp/.local/lib/python3.5/site-packages/torch/nn/parallel/scatter_gather.py", line 62, in gather_map
    return type(out)(map(gather_map, zip(*outputs)))
  File "/home/michaelp/.local/lib/python3.5/site-packages/torch/nn/parallel/scatter_gather.py", line 54, in gather_map
    return Gather.apply(target_device, dim, *outputs)
  File "/home/michaelp/.local/lib/python3.5/site-packages/torch/nn/parallel/_functions.py", line 65, in forward
    return comm.gather(inputs, ctx.dim, ctx.target_device)
  File "/home/michaelp/.local/lib/python3.5/site-packages/torch/cuda/comm.py", line 160, in gather
    return torch._C._gather(tensors, dim, destination)
TypeError: _gather(): incompatible function arguments. The following argument types are supported:
    1. (tensors: List[at::Tensor], dim: int, destination_index: Optional[int]) -> at::Tensor

Invoked with: (tensor([[[-0.0603, -0.0166, -0.0857,  0.0781,  0.1473,  0.1543,  0.0239,
           0.0820,  0.0887, -0.0435]]],
       device='cuda:0', grad_fn=<CudnnRnnBackward>),), 0, device(type='cuda')
```

## System Info

- PyTorch or Caffe2: PyTorch
- How you installed PyTorch (conda, pip, source): pip
- PyTorch version: 0.4.1
- Python version: 3.5.2

### PR DESCRIPTION
**Pull request by weiyangfb**
- fixes #9984

### CODE REVIEW

**Comment by weiyangfb** [comment]
@vishwakftw I see, I will change them as well

File: test/test_nn.py [hunk 1]
```diff
@@ -2810,6 +2810,12 @@ def test_data_parallel_small_back(self):
         out = dp.data_parallel(l, i, (0, 1))
         self.assertEqual(out, l(i))
 
+        # test output_device
+        l = nn.Linear(10, 5).float().cuda()
+        i = Variable(torch.randn(20, 10).float().cuda())
+        out = dp.data_parallel(l, i, (0, 1), torch.device('cuda'))
```
Comment (ailzhang) [inline test/test_nn.py]: This test should also be moved into test_distributed.py I think

Comment (weiyangfb) [reply test/test_nn.py]: This is testing data_parallel (single machine multiple gpus), so should be fine at test_nn? I will put DistributedDataParallel related tests to test_distributed

**Review by li-roy — APPROVED** [review]
This looks good to me.

**Comment by li-roy** [comment]
actually, can we add a test for the other two code paths as well?

**Review by ssnl — CHANGES_REQUESTED** [review]
The better fix is to make `scatter`, `gather`, `parallel_apply`, etc. to accept device objects (vs. converting to idx in DP). You can also make `device_ids` to support device objects this way.

**Comment by fmassa** [comment]
ping @weiyangfb on @SsnL suggestion.

**Comment by SsnL** [comment]
you can probably use/adapt `torch.cuda._get_device_index` now to do that after #10833 .

**Comment by weiyangfb** [comment]
is this good to go? @SsnL

**Comment by weiyangfb** [comment]
is this good to go? @SsnL @teng-li @ailzhang

File: torch/nn/parallel/parallel_apply.py [hunk 2]
```diff
@@ -36,7 +37,7 @@ def parallel_apply(modules, inputs, kwargs_tup=None, devices=None):
         assert len(modules) == len(devices)
     else:
         devices = [None] * len(modules)
-
+    devices = list(map(lambda x: _get_device_index(x, True), devices))
```
Comment (teng-li) [inline torch/nn/parallel/parallel_apply.py]: Please add a comment for devices in line 26 as well on the expected types

**Comment by weiyangfb** [comment]
@SsnL would you like to take quick pass on this? The updates are separated test function and doc fixes. Thanks!

**Review by ssnl — APPROVED** [review]
lgtm, but there is one remaining nit to be addressed

**Comment by weiyangfb** [comment]
@SsnL I see, fixed more places with `Device` -> `torch.device`

**Pull request closed without merging**